Added volume buttons to media_player more-info

rmkraus opened the pull request:
Added volume buttons to the media_player more-info panel. The volume
buttons appear if the supportsVolumeButtons flag is turned on and the
player is not off. They send a volume up/down service call every half
second while they are being pressed.

<img width="275" alt="screen shot 2016-01-11 at 12 02 12 am" src="https://cloud.githubusercontent.com/assets/4956442/12227220/d098dc10-b7f6-11e5-8bdb-219c3a700836.png">

rmkraus committed:
Merge remote-tracking branch 'balloob/master'

rmkraus committed:
Added volume buttons to media_player more-info

Added volume buttons to the media_player more-info panel. The volume
buttons appear if the supportsVolumeButtons flag is turned on and the
player is not off. They send a volume up/down service call every half
second while they are being pressed.

rmkraus committed:
Fixing lint errors and mousedown on mobile browsers

1) Fixed a handful of errors caught by the linter.
2) Mobile browsers register a press and hold as a touch event rather
than a regular mouse down event. The code has been updated so that
these touch events will be registered and sent to the mouse down
handler.

rmkraus committed:
Fixing more linter errors

Changing lets to consts

rmkraus commented:
Before I head out for the night, I just remembered this:

The mute icon does not change dependant upon isMuted like normal. This is because it would become too confusing next to the other volume buttons. However, the volume down and up buttons do gray out and disable themselves when the volume is muted.

rmkraus committed:
Removed debugging line

Removed a debugging line form media_player more-info code.

rmkraus committed:
Hide play button in media player if there is not media to control

If the media player does not have media to control, the
play/pause/stop, previous, and next buttons will be hidden. In order
for the playback buttons to be shown, the status of the media player
must be playing, paused, or unknown. This way, if the status is On, the
playback buttons will be hidden.

balloob commented on src/more-infos/more-info-media_player.js:
```diff
@@ -136,6 +167,24 @@ export default new Polymer({
+    if (force || (obj !== undefined && obj.pointerDown)) {
+      this.callService(service);
+      const _this = this;
+      setTimeout(function callback() {_this.handleVolumeWorker(service, obj, false);}, 500);
```
```javascript
this.async(() => this.handleVolumeWorker(service, obj, false), 500);
```

balloob commented on src/more-infos/more-info-media_player.js:
```diff
@@ -136,6 +167,24 @@ export default new Polymer({
+  },
+
+  handleVolumeWorker(service, obj, force) {
+    if (force || (obj !== undefined && obj.pointerDown)) {
```
What's up with the pointerDown check?

rmkraus replied:
It allows the function to loop while the button is being held down. The first time in the function, it is forced to run. It then runs every 0.5 seconds until the button is released.

balloob replied:
What confuses me is that pointerDown is not part of the official HTML spec…

rmkraus replied:
https://elements.polymer-project.org/elements/paper-icon-button#pointerDown

That is where I got it from.

balloob replied:
Ooh it's a polymer thing, excellent. My bad

balloob commented on src/more-infos/more-info-media_player.js:
```diff
@@ -70,12 +70,34 @@ export default new Polymer({
+  attached() {
+    // This is required to bind a mousedown event in all browsers
+    const _this = this;
+    this.$.volumeUp.onmousedown = function onVolumeUpDown() {_this.handleVolumeUp();};
```
If you want to use JavaScript to attach event listeners, use this approach:

```javascript
this.$.volumeUp.addEventListener('mousedown', () => this.handleVolumeUp());
```

That being said - you can just use the template to attach event listeners, that is the preferred method:

```html
<paper-icon-button id="volumeDown" disabled$='[[isMuted]]'
  icon="mdi:volume-medium" on-mousedown="handleVolumeUp"></paper-icon-button>
```

That being said 2 - we don't have any code within Home Assistant that currently works by keeping it pressed and it is seldom seen in touch UIs. We shouldn't do this.

rmkraus replied:
In my opinion, this is a crucial feature for volume up and down buttons. Volume buttons become impossible to use if you have to click it for each increment you want to move.

If I want to move the volume from 20 to 35 (a very regular request), clicking the increment button 15 times gets crazy. Button mashing is far from accurate and is quite fatiguing.

I can definitely change the approach. JavaScript is not my strong suit. I bound the events in the JavaScript because, when I used the preferred method, I was having trouble in Safari. I can try the preferred method again. I may have been doing something else wrong.

rmkraus committed:
Clean up the media player more-info code

1) Event binding was moved to the html template as is tradition.
2) setTimeout call was replaced with this.async… as is tradition.

rmkraus commented:
We should be good here as long as you'll let me have the button repeat as it is held down.
